lib/cudadrv/execution.jl

```diff
@@ -141,7 +141,11 @@ end
         converted_args[i] = gensym()
         arg_ptrs[i] = gensym()
         push!(ex.args, :($(converted_args[i]) = Base.cconvert($(types[i]), args[$i])))
-        push!(ex.args, :($(arg_ptrs[i]) = Base.unsafe_convert($(types[i]), $(converted_args[i]))))
+        if types[i] === Symbol
+            push!(ex.args, :($(arg_ptrs[i]) = Base.pointer_from_objref($(converted_args[i]))))
```

Comment:
That seems like a hack... but I'm not sure how to generalize. Maybe by keeping the type `Symbol`? How do you pass symbol references to C using `ccall`?

Reply:
`ccall` as a fun/weird ABI for Any and therefore Symbol. Where it just passes the object pointer as a ptr...

Emulating that behavior has been an issues as well for Enzyme. (Sometimes `pointer_from_objref` decides not to work.)

The other solution is to not wrap types with pointer identity in `Ref`.

Reply:
E.g. this issue was as far as I could tell here https://github.com/JuliaGPU/CUDA.jl/blob/45c7659318bbac9c39572dd99022ac9c6541a80c/lib/cudadrv/execution.jl#L21-L29

```
julia> r_sym = Base.RefValue(:var)
Base.RefValue{Symbol}(:var)

julia> Base.unsafe_convert(Ptr{Cvoid}, r_sym)
Ptr{Nothing} @0x000078632a27bf00

julia> Base.pointer_from_objref(:var)
Ptr{Nothing} @0x000078632a27bf00
```

And we want it to be a pointer to that pointer.

Reply:
> `ccall` as a fun/weird ABI for Any and therefore Symbol. Where it just passes the object pointer as a ptr...

Why not copy that? Seems sensible to pass things that are by-reference by their reference.

Reply:
Oh, absolutely. The problem here is the semantics of `Ref`. We might need a private implementation that handles this correctly. In https://github.com/JuliaGPU/CUDA.jl/blob/45c7659318bbac9c39572dd99022ac9c6541a80c/lib/cudadrv/execution.jl#L21-L29 we want a pointer to a pointer and `Ref` gives us the pointer of the object.

Reply:
Why do we need a Ref? `ccall(:jl_, Nothing, (Any,), :test)` works, no?

Reply:
Because the argument buffer for the CUDA call consits out of pointers to data to copy IIUC.

Reply:
Let's see if 1ce2219c561c16278ae108e2ceb80767a43ed35e does the trick.

Reply:
Oh OK, I thought you were suggesting to use Ref as input to the `cudacall`; this is just to _implement_ those semantics, got it.